### PR TITLE
[CONTINT-4112] Add PodWatcher in the autoscaling controller

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
This PR:
- Adds resource information to the KubernetesPod objects collected by the kubeapiserver collector
- Adds a PodWatcher, indexing pods for the autoscaling controller

### Motivation
- The autoscaling controller needs to retrieve pods per pod owner.

### Additional Notes
We could potentially pre-aggregate pod count per owner + status and pod count per owner + requests/limits.

### Possible Drawbacks / Trade-offs
Memory usage will increase when the cluster-tagger or the autoscaling controller is enabled.

### Describe how to test/QA your changes

- Workloadmeta changes:
Deploy the Cluster-Agent with `autoscaling.workload.enabled`. Make sure to visualize the pods in the output of `agent workload-list -v`. 

- Pod Watcher:
It is not used yet.

### CODE REVIEW

File: comp/core/workloadmeta/collectors/internal/kubeapiserver/pod.go [hunk 1]
```diff
@@ -73,6 +73,26 @@ func newPodParser(annotationsExclude []string) (objectParser, error) {
...
+func convertResource(c corev1.ResourceList) map[string]string {
+	res := make(map[string]string, len(c))
+	for k, v := range c {
+		res[string(k)] = v.String()
+	}
+	return res
+}
+
+func containersToOrchestratorContainers(cont []corev1.Container) []workloadmeta.OrchestratorContainer {
+	containers := make([]workloadmeta.OrchestratorContainer, 0, len(cont))
```
Comment (sblumenthal) [inline comp/core/workloadmeta/collectors/internal/kubeapiserver/pod.go:85]: For now, does it make sense to short-circuit this function and have it return an empty list or nil if `cfg.GetBool("autoscaling.workload.enabled")` is not true?

Comment (AliDatadog) [reply comp/core/workloadmeta/collectors/internal/kubeapiserver/pod.go:85]: I think it would absolutely make sense

Comment (AliDatadog) [reply comp/core/workloadmeta/collectors/internal/kubeapiserver/pod.go:91]: Done in the last commit

File: pkg/clusteragent/autoscaling/workload/pod_watcher.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	case workloadmeta.EventTypeUnset:
+		pw.handleUnsetEvent(event)
+	default:
+		log.Errorf("Ignoring event type %d", event.Type)
+	}
+	pw.mutex.Unlock()
+}
+
+func (pw *podWatcher) handleSetEvent(event workloadmeta.Event) {
+	pod, ok := event.Entity.(*workloadmeta.KubernetesPod)
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/pod_watcher.go:114]: Common code that could be in `handleEvent`.

File: pkg/clusteragent/autoscaling/workload/pod_watcher.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		return
+	}
+	delete(pw.podsPerPodOwner[podOwner], pod.ID)
+	if len(pw.podsPerPodOwner[podOwner]) == 0 {
+		delete(pw.podsPerPodOwner, podOwner)
+	}
+}
+
+func getNamespacedPodOwner(ns string, owner *workloadmeta.KubernetesPodOwner) NamespacedPodOwner {
+	res := NamespacedPodOwner{
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/pod_watcher.go:153]: Code is slightly incorrect. `ReplicaSet` can be created on their own, we try to guess that in `ParseDeploymentForReplicaSet`, that can return empty.

File: pkg/clusteragent/autoscaling/workload/pod_watcher.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	pw.mutex.Lock()
+	switch event.Type {
+	case workloadmeta.EventTypeSet:
+		pw.handleSetEvent(event)
+	case workloadmeta.EventTypeUnset:
+		pw.handleUnsetEvent(event)
+	default:
+		log.Errorf("Ignoring event type %d", event.Type)
+	}
+	pw.mutex.Unlock()
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/workload/pod_watcher.go:110]: `defer` seems appropriate here.

File: pkg/clusteragent/autoscaling/workload/pod_watcher.go (hunk 3 above)
Comment (AliDatadog) [reply pkg/clusteragent/autoscaling/workload/pod_watcher.go:148]: Should we support pods without deployments ?

Comment (vboulineau) [reply pkg/clusteragent/autoscaling/workload/pod_watcher.go:148]: Yes for horizontal we support anything.

**Review by vboulineau — APPROVED** [review]

**Review by davidor — APPROVED** [review]
:+1: for container-platform files

**Pull request merged**